# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/DirectoryMetadata.cpp
# Commit: a576f5346c6c
# Full Hash: a576f5346c6cfc32a5b147158e48ec1f4fcf2a8f
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2025-06-12 16:31:09
# Regressor Bug: 1962133
# File Overlap Count: 5
# Description:
#   Bug 1962133 - QM: Prepare metadata header infrastructure for accessed flag addition and reduce duplication; r=dom-storage-reviewers,hsingh
#   
#   This patch restructures metadata header reading and writing code to reduce
#   duplication and establish clean infrastructure for adding the accessed flag to
#   the metadata file header.
# ==============================================================================

diff -r f4284b0494fc -r a576f5346c6c dom/quota/DirectoryMetadata.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/quota/DirectoryMetadata.cpp	Wed Jun 11 20:09:11 2025 +0000
@@ -0,0 +1,102 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "DirectoryMetadata.h"
+
+#include "mozilla/Result.h"
+#include "mozilla/dom/quota/Assertions.h"
+#include "mozilla/dom/quota/CommonMetadata.h"
+#include "mozilla/dom/quota/QuotaCommon.h"
+#include "mozilla/dom/quota/ResultExtensions.h"
+#include "mozilla/dom/quota/StreamUtils.h"
+#include "nsIBinaryInputStream.h"
+#include "nsIBinaryOutputStream.h"
+
+namespace mozilla::dom::quota {
+
+Result<OriginStateMetadata, nsresult> ReadDirectoryMetadataHeader(
+    nsIBinaryInputStream& aStream) {
+  AssertIsOnIOThread();
+
+  OriginStateMetadata originStateMetadata;
+
+  QM_TRY_UNWRAP(originStateMetadata.mLastAccessTime,
+                MOZ_TO_RESULT_INVOKE_MEMBER(aStream, Read64));
+
+  QM_TRY_UNWRAP(originStateMetadata.mPersisted,
+                MOZ_TO_RESULT_INVOKE_MEMBER(aStream, ReadBoolean));
+
+  QM_TRY_INSPECT(const bool& reservedData1,
+                 MOZ_TO_RESULT_INVOKE_MEMBER(aStream, Read32));
+  if (reservedData1 != 0) {
+    QM_TRY(MOZ_TO_RESULT(false));
+  }
+
+  // XXX Use for the persistence type.
+  QM_TRY_INSPECT(const bool& reservedData2,
+                 MOZ_TO_RESULT_INVOKE_MEMBER(aStream, Read32));
+  Unused << reservedData2;
+
+  return originStateMetadata;
+}
+
+nsresult WriteDirectoryMetadataHeader(
+    nsIBinaryOutputStream& aStream,
+    const OriginStateMetadata& aOriginStateMetadata) {
+  AssertIsOnIOThread();
+
+  QM_TRY(MOZ_TO_RESULT(aStream.Write64(aOriginStateMetadata.mLastAccessTime)));
+
+  QM_TRY(MOZ_TO_RESULT(aStream.WriteBoolean(aOriginStateMetadata.mPersisted)));
+
+  // Reserved data 1
+  QM_TRY(MOZ_TO_RESULT(aStream.Write32(0)));
+
+  // Reserved data 2
+  QM_TRY(MOZ_TO_RESULT(aStream.Write32(0)));
+
+  return NS_OK;
+}
+
+Result<OriginStateMetadata, nsresult> LoadDirectoryMetadataHeader(
+    nsIFile& aDirectory) {
+  AssertIsOnIOThread();
+
+  QM_TRY_INSPECT(
+      const auto& stream,
+      GetBinaryInputStream(aDirectory, nsLiteralString(METADATA_V2_FILE_NAME)));
+
+  QM_TRY_INSPECT(const OriginStateMetadata& originStateMetadata,
+                 ReadDirectoryMetadataHeader(*stream));
+
+  QM_TRY(MOZ_TO_RESULT(stream->Close()));
+
+  return originStateMetadata;
+}
+
+nsresult SaveDirectoryMetadataHeader(
+    nsIFile& aDirectory, const OriginStateMetadata& aOriginStateMetadata) {
+  AssertIsOnIOThread();
+
+  QM_TRY_INSPECT(
+      const auto& file,
+      CloneFileAndAppend(aDirectory, nsLiteralString(METADATA_V2_FILE_NAME)));
+
+  QM_TRY_INSPECT(const auto& stream,
+                 GetBinaryOutputStream(*file, FileFlag::Update));
+  MOZ_ASSERT(stream);
+
+  QM_TRY(MOZ_TO_RESULT(
+      WriteDirectoryMetadataHeader(*stream, aOriginStateMetadata)));
+
+  QM_TRY(MOZ_TO_RESULT(stream->Flush()));
+
+  QM_TRY(MOZ_TO_RESULT(stream->Close()));
+
+  return NS_OK;
+}
+
+}  // namespace mozilla::dom::quota